bdai_ros2_wrappers/bdai_ros2_wrappers/logging.py

```diff
@@ -37,6 +44,229 @@
+
+    Args:
+        name: logger name.
+        level: log severity level.
+        origin: log call site.
+        throttle_duration_sec: optional period for throttling, in seconds.
+        throttle_time_source: optional time source for throttling, defaults to system time.
+        skip_first: if True, skip the first log.
+        once: if True, only log once.
+
+    Returns:
+        a logging function.
+    """
+
+    def log(message: str) -> bool:
+        impl.rclpy_logging_rcutils_log(
+            level,
+            name,
+            message,
+            origin.function,
+            origin.filename,
+            origin.lineno,
+        )
+        return True
+
+    if throttle_duration_sec is not None:
+        throttle_period = Duration(
+            seconds=throttle_duration_sec,
+        )
+        log = throttle(
+            log,
+            min_period=throttle_period,
+            time_source=throttle_time_source,
+            fill_value=False,
+        )
+    if skip_first:
+        log = skip(log, num_times=1, fill_value=False)
+    if once:
+        log = cap(log, num_times=1, fill_value=False)
+    return log
+
+
+class MemoizingRcutilsLogger:
+    """An alternative, more efficient implementation of RcutilsLogger.
+
+    MemoizingRcutilsLogger caches logging call configuration to speed up
+    subsequent invocations.
+    """
+
+    def __init__(self, name: str = "") -> None:
+        """Initializes the logger with the given `name`, or the root logger if none is provided."""
+        self.name = name
+
+    def get_child(self, name: str) -> "MemoizingRcutilsLogger":
+        """Gets a child logger with the given `name`."""
+        if not name:
+            raise ValueError("Need name for child logger.")
+        if self.name:
+            name = self.name + "." + name
+        return MemoizingRcutilsLogger(name)
+
+    def set_level(self, level: typing.Union[int, LoggingSeverity]) -> None:
+        """Sets logger severity `level`."""
+        impl.rclpy_logging_set_logger_level(self.name, LoggingSeverity(level))
+
+    def get_effective_level(self) -> LoggingSeverity:
+        """Gets the effective logger severity level.
+
+        The effective severity level of a logger is the first severity level set in the logger
+        genealogy (ie. its own or that of its parent or that of its grandparent and so on), or
+        the default severity level when no severity level is.
+        """
+        return LoggingSeverity(impl.rclpy_logging_get_logger_effective_level(self.name))
+
+    def is_enabled_for(self, level: typing.Union[int, LoggingSeverity]) -> bool:
+        """Checks whether the logger is enabled for logs at the given severity `level`."""
+        return impl.rclpy_logging_logger_is_enabled_for(self.name, LoggingSeverity(level))
+
+    def log(
+        self,
+        message: str,
+        level: typing.Union[int, LoggingSeverity],
+        origin: typing.Optional[inspect.Traceback] = None,
```

Comment:
`type | None` could be used in replacement of `Optional`

Reply:
@robodreamer `ros_utilities` supports python 3.8 which doesn't support the pipe operator for types.